### PR TITLE
Support TidbMonitor Ingress

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
Support TidbMonitor Ingress

Close https://github.com/pingcap/tidb-operator/issues/2276

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Support `TidbMonitor` to configure `Ingress`
```

### CODE REVIEW

File: pkg/controller/generic_control.go [hunk 1]
```diff
@@ -301,6 +304,39 @@ func (w *typedWrapper) CreateOrUpdateService(controller runtime.Object, svc *cor
...
+func (w *typedWrapper) CreateOrUpdateIngress(controller runtime.Object, ingress *extensionsv1beta1.Ingress) (*extensionsv1beta1.Ingress, error) {
+	result, err := w.GenericControlInterface.CreateOrUpdate(controller, ingress, func(existing, desired runtime.Object) error {
+		existingIngress := existing.(*extensionsv1beta1.Ingress)
+		desiredIngress := desired.(*extensionsv1beta1.Ingress)
+
+		if existingIngress.Annotations == nil {
+			existingIngress.Annotations = map[string]string{}
+		}
+		for k, v := range desiredIngress.Annotations {
+			desiredIngress.Annotations[k] = v
```
Comment (cofyc) [inline pkg/controller/generic_control.go:316]: existingIngress

File: pkg/monitor/monitor/monitor_manager.go [hunk 2]
```diff
@@ -272,3 +285,44 @@ func (mm *MonitorManager) syncTidbMonitorRbac(monitor *v1alpha1.TidbMonitor) (*c
...
+		return mm.removeIngressIfExist(monitor, grafanaName(monitor))
+	} else if monitor.Spec.Grafana.Ingress == nil {
+		return mm.removeIngressIfExist(monitor, grafanaName(monitor))
+	}
+	ingress := getGrafanaIngress(monitor)
+	_, err := mm.typedControl.CreateOrUpdateIngress(monitor, ingress)
+	return err
+}
+
+// If the dashboard or dashboard ingress is nil when ingress also existed, we would remove Ingress
```
Comment (cofyc) [inline pkg/monitor/monitor/monitor_manager.go:318]:
```suggestion
// removeIngressIfExist removes Ingress if it exists
```

you didn't check any spec is nil in this function

File: pkg/monitor/monitor/monitor_manager.go [hunk 3]
```diff
@@ -272,3 +285,44 @@ func (mm *MonitorManager) syncTidbMonitorRbac(monitor *v1alpha1.TidbMonitor) (*c
...
+	_, err := mm.typedControl.CreateOrUpdateIngress(monitor, ingress)
+	return err
+}
+
+func (mm *MonitorManager) syncGrafanaIngress(monitor *v1alpha1.TidbMonitor) error {
+	if monitor.Spec.Grafana == nil {
+		return mm.removeIngressIfExist(monitor, grafanaName(monitor))
+	} else if monitor.Spec.Grafana.Ingress == nil {
+		return mm.removeIngressIfExist(monitor, grafanaName(monitor))
+	}
```
Comment (cofyc) [inline pkg/monitor/monitor/monitor_manager.go:312]:
```suggestion
	if monitor.Spec.Grafana == nil || monitor.Spec.Grafana.Ingress == nil  {
		return mm.removeIngressIfExist(monitor, grafanaName(monitor))
	}
```

shorter

File: pkg/controller/generic_control.go (hunk 1 above)
Comment (Yisaer) [reply pkg/controller/generic_control.go]: updated.

File: pkg/monitor/monitor/monitor_manager.go [hunk 4]
```diff
@@ -272,3 +285,44 @@ func (mm *MonitorManager) syncTidbMonitorRbac(monitor *v1alpha1.TidbMonitor) (*c
...
+		return err
+	}
+
+	return mm.syncGrafanaIngress(monitor)
+}
+
+func (mm *MonitorManager) syncPrometheusIngress(monitor *v1alpha1.TidbMonitor) error {
+	if monitor.Spec.Prometheus.Ingress == nil {
+		return mm.removeIngressIfExist(monitor, prometheusName(monitor))
+	}
```
Comment (cofyc) [inline pkg/monitor/monitor/monitor_manager.go:300]: should we check `monitor.Spec.Prometheus` too like grafana?

Comment (Yisaer) [reply pkg/monitor/monitor/monitor_manager.go:300]: We don't have to check as `Prometheus` is a non-nil in `TidbMonitorSpec`

File: pkg/controller/generic_control.go [hunk 5]
```diff
@@ -301,6 +304,39 @@ func (w *typedWrapper) CreateOrUpdateService(controller runtime.Object, svc *cor
...
+		for k, v := range desiredIngress.Annotations {
+			existingIngress.Annotations[k] = v
+		}
+		existingIngress.Labels = desiredIngress.Labels
+		equal, err := IngressEqual(desiredIngress, existingIngress)
+		if err != nil {
+			return err
+		}
+		if !equal {
+			// record desiredSvc Spec in annotations in favor of future equality checks
```
Comment (cofyc) [inline pkg/controller/generic_control.go:324]:
```suggestion
			// record desiredIngress Spec in annotations in favor of future equality checks
```

**Pull request merged**